Src/VimWpf/Implementation/ImeCoordinator/ImeCoordinator.cs

```diff
@@ -0,0 +1,121 @@
+﻿using System;
+using System.ComponentModel.Composition;
+using System.Windows.Input;
+using Microsoft.VisualStudio.Text.Editor;
+
+namespace Vim.UI.Wpf.Implementation.ImeCoordinator
+{
+    [Export(typeof(IVimBufferCreationListener))]
+    internal sealed class ImeCoordinator : IVimBufferCreationListener
+    {
+        private readonly IVim _vim;
+
+        private bool _inInputMode;
+        private InputMethodState _lastInputMethodState;
+
+        [ImportingConstructor]
+        internal ImeCoordinator(IVim vim)
+        {
+            _vim = vim;
+
+            _inInputMode = true;
+            _lastInputMethodState = InputMethodState.DoNotCare;
+        }
+
+        void VimBufferCreated(IVimBuffer vimBuffer)
+        {
+            if (vimBuffer.TextView is IWpfTextView textView)
+            {
+                vimBuffer.KeyInputProcessed += OnKeyInputProcessed;
+                textView.GotAggregateFocus += OnGotFocus;
+                vimBuffer.Closed += OnBufferClosed;
+            }
+        }
+
+        private void OnKeyInputProcessed(object sender, KeyInputProcessedEventArgs e)
+        {
+            if (sender is IVimBuffer vimBuffer)
+            {
+                OnInputModeRelatedEvent(vimBuffer);
+            }
+        }
+
+        private void OnGotFocus(object sender, EventArgs e)
+        {
+            if (sender is IWpfTextView textView)
+            {
+                if (_vim.TryGetVimBuffer(textView, out IVimBuffer vimBuffer))
+                {
+                    OnInputModeRelatedEvent(vimBuffer);
+                }
+            }
+        }
+
+        private void OnBufferClosed(object sender, EventArgs e)
+        {
+            if (sender is IVimBuffer vimBuffer && vimBuffer.TextView is IWpfTextView textView)
+            {
+                vimBuffer.KeyInputProcessed -= OnKeyInputProcessed;
+                textView.GotAggregateFocus -= OnGotFocus;
+                vimBuffer.Closed -= OnBufferClosed;
+            }
+        }
+
+        private void OnInputModeRelatedEvent(IVimBuffer vimBuffer)
+        {
+            var wasInInputMode = _inInputMode;
+            var isInInputMode = IsInputMode(vimBuffer);
+            _inInputMode = isInInputMode;
+
+            if (wasInInputMode != isInInputMode)

```

Comment:
Imagine there are two buffers both in insert mode with `ImeState` set to `InputMethodState.On`. Now I hit escape in buffer 1, switch to buffer 2 and hit escape. Won't we end up with with both `_lastInputMethodState` and `ImeState.Current` set to `Off`?